photutils/isophote/ellipse_model.pyx

```diff
@@ -0,0 +1,157 @@
+# Licensed under a 3-clause BSD style license - see LICENSE.rst
+# cython: language_level=3
+"""
+Faster evaluation of ellipses from model.py.
+"""
+
+import cython
+import numpy as np
+
+cimport numpy as cnp
+
+__all__ = ['build_ellipse_model_c']
+
+cnp.import_array()
+
+cdef extern from "math.h":
+
+    double cos(double x)
+    double sin(double x)
+    double sqrt(double x)
+
+
+DTYPE = np.float64
+ctypedef cnp.float64_t DTYPE_t
+
+cdef inline double get_intens_no_harmonics(double intens0, double phi, double a3, double b3, double a4, double b4):
+    return intens0
+
+cdef inline double get_intens_harmonics(double intens0, double phi, double a3, double b3, double a4, double b4):
+    return (
+        intens0
+        + a3 * sin(3.0 * phi)
+        + b3 * cos(3.0 * phi)
+        + a4 * sin(4.0 * phi)
+        + b4 * cos(4.0 * phi)
+    )
+
+
+def build_ellipse_model_c(
+    unsigned int n_rows,
+    unsigned int n_cols,
+    double phi_min,
+    double phi_max,
+    cnp.ndarray[DTYPE_t, ndim=1] finely_spaced_sma,
+    cnp.ndarray[DTYPE_t, ndim=1] intens_array,
+    cnp.ndarray[DTYPE_t, ndim=1] eps_array,
+    cnp.ndarray[DTYPE_t, ndim=1] pa_array,
+    cnp.ndarray[DTYPE_t, ndim=1] x0_array,
+    cnp.ndarray[DTYPE_t, ndim=1] y0_array,
+    cnp.ndarray[DTYPE_t, ndim=1] a3_array = None,
+    cnp.ndarray[DTYPE_t, ndim=1] b3_array = None,
+    cnp.ndarray[DTYPE_t, ndim=1] a4_array = None,
+    cnp.ndarray[DTYPE_t, ndim=1] b4_array = None,
+):
+    cdef cython.Py_ssize_t len_sma
+    len_sma = len(finely_spaced_sma)
+    for array in (intens_array, eps_array, pa_array, x0_array, y0_array):
+        if len(array) != len_sma:
+            raise ValueError(f"All input arrays must be same length={len_sma}")
+    harmonics_is_none = (
+        array is None for array in (a3_array, b3_array, a4_array, b4_array)
+    )
```

Comment:
```suggestion
    harmonics_arrays = (a3_array, b3_array, a4_array, b4_array)
    harmonics_is_none = [array is None for array in harmonics_arrays]
```

As a generator, I think the `all(harmonics_is_none)` can short circuit leaving the generator partially consumed, and the `if any(harmonics_is_none)` line will start from the partially consumed generator.  It's better to define `harmonics_is_none` as a list.  This also avoids the generator possibly being calculated twice (once for `all` and once for `any`).